core/Pos/Crypto/HD.hs

```diff
@@ -77,25 +84,46 @@ deriveHDSecretKey passPhrase (SecretKey xprv) childIndex
   | otherwise =
       SecretKey $ deriveXPrv passPhrase xprv (childIndex - maxHardened - 1)
 
+addrAttrNonce :: ByteString
+addrAttrNonce = "serokellfore"
+
+addrAttrHeader :: ByteString
+addrAttrHeader = "addressattr"
+
 -- | Serialize tree path and encrypt it using passphrase via ChaChaPoly1305.
 packHDAddressAttr :: HDPassphrase -> [Word32] -> HDAddressPayload
-packHDAddressAttr (HDPassphrase simkey) path = do
-    let pathSer = encodeStrict path
-    let packCF =
+packHDAddressAttr (HDPassphrase passphrase) path = do
+    let !pathSer = encodeStrict path
+    let !packCF =
           encryptChaChaPoly
-              "serokellfore"
-              simkey
-              "addressattr" -- I don't know what does mean this parameter.
-                            -- Seems value isn't important...
+              addrAttrNonce
+              passphrase
+              addrAttrHeader -- I don't know what does this parameter mean. Seems value isn't important...
               pathSer
     case packCF of
-        CryptoFailed er -> error $ "Error during packHDAddressAttr: " <> show er
+        CryptoFailed er -> error $ "Error in packHDAddressAttr: " <> show er
         CryptoPassed p  -> HDAddressPayload p
 
+unpackHDAddressAttr :: MonadFail m => HDPassphrase -> HDAddressPayload -> m [Word32]
+unpackHDAddressAttr (HDPassphrase passphrase) (HDAddressPayload payload) = do
+    let !unpackCF =
+          decryptChaChaPoly
+              addrAttrNonce
+              passphrase
+              addrAttrHeader
+              payload
+    case unpackCF of
+        Left er ->
+            fail $ "Error in unpackHDAddressAttr, during decryption: " <> show er
+        Right p -> case decodeFull . BSL.fromStrict $ p of
+            Left er ->
+                fail $ "Error in unpackHDAddressAttr, during deserialization: " <> show er
+            Right path -> pure path
+
 -- Wrapper around ChaChaPoly1305 module.
 encryptChaChaPoly
     :: ByteString -- nonce (12 random bytes)
-    -> ByteString -- symmetric key
+    -> ByteString -- symmetric key (must be 32 bytes)
     -> ByteString -- optional associated data (won't be encrypted)
```

Comment:
Please add comment that header is chunk of data we want to transfer unecncrypted but still want it to be part of tag digest (so tag verifies validity of both encrypted data and unencrypted header)